packages/react-components/react-provider/src/components/FluentProvider/useFluentProviderThemeStyleTag.ts

```diff
@@ -1,63 +1,67 @@
+  : useIsomorphicLayoutEffect;
+
+const createStyleTag = (target: Document | undefined, id: string) => {
+  if (!target) {
+    return undefined;
+  }
+  const tag = target.createElement('style');
+  tag.setAttribute('id', id);
+  target.head.appendChild(tag);
+  return tag;
+};
+
+const insertSheet = (tag: HTMLStyleElement, rule: string) => {
+  const sheet = tag.sheet;
+
+  if (sheet) {
+    if (sheet.cssRules.length > 0) {
+      sheet.deleteRule(0);
+    }
+    sheet.insertRule(rule, 0);
+  } else if (process.env.NODE_ENV !== 'production') {
+    // eslint-disable-next-line no-console
+    console.error('FluentProvider: No sheet available on styleTag, styles will not be inserted into DOM.');
+  }
+};
+
 /**
  * Writes a theme as css variables in a style tag on the provided targetDocument as a rule applied to a CSS class
  *
  * @returns CSS class to apply the rule
  */
 export const useFluentProviderThemeStyleTag = (options: Pick<FluentProviderState, 'theme' | 'targetDocument'>) => {
   const { targetDocument, theme } = options;
+  const styleTag = React.useRef<HTMLStyleElement>();
 
   const styleTagId = useId(fluentProviderClassNames.root);
-  const styleTag = React.useMemo(() => {
-    if (!targetDocument) {
-      return null;
-    }
-
-    const tag = targetDocument.createElement('style');
-    tag.setAttribute('id', styleTagId);
-    targetDocument.head.appendChild(tag);
-    return tag;
-  }, [styleTagId, targetDocument]);
 
-  const cssRule = React.useMemo(() => {
-    const cssVarsAsString = theme
+  const cssVarsAsString = React.useMemo(() => {
+    return theme
       ? (Object.keys(theme) as (keyof typeof theme)[]).reduce((cssVarRule, cssVar) => {
           cssVarRule += `--${cssVar}: ${theme[cssVar]}; `;
           return cssVarRule;
         }, '')
       : '';
+  }, [theme]);
 
-    // result: .fluent-provider1 { --css-var: '#fff' }
-    return `.${styleTagId} { ${cssVarsAsString} }`;
-  }, [theme, styleTagId]);
-  const previousCssRule = usePrevious(cssRule);
-
-  if (styleTag && previousCssRule !== cssRule) {
-    const sheet = styleTag.sheet;
-
-    if (sheet) {
-      if (sheet.cssRules.length > 0) {
-        sheet.deleteRule(0);
-      }
-      sheet.insertRule(cssRule, 0);
-    } else if (process.env.NODE_ENV !== 'production') {
-      // eslint-disable-next-line no-console
-      console.error('FluentProvider: No sheet available on styleTag, styles will not be inserted into DOM.');
-    }
-  }
+  const rule = `.${styleTagId} { ${cssVarsAsString} }`;
 
-  // Removes the style tag from the targetDocument on unmount or change
-  React.useEffect(() => {
-    return () => {
-      if (styleTag) {
-        styleTag.remove();
-      }
-    };
-  }, [styleTag]);
+  useInsertionEffect(() => {
+    styleTag.current = createStyleTag(targetDocument, styleTagId);
+    if (styleTag.current) {
+      insertSheet(styleTag.current, rule);
+      return () => {
+        styleTag.current?.remove();
```

Comment:
```suggestion
        styleTag.current.remove();
```

We are checking that `styleTag.current` is defined few lines above so it's not needed, right?

Reply:
i thought so as well, but linting was throwing an error

Reply:
I think the error was because it's a ref, so styleTag.current could theoretically be changed/nulled by other code before the cleanup function runs. In this case, you're not using the styleTag ref outside of the effect anyways, so you could delete the ref and make it a local const instead.

```ts
  useInsertionEffect(() => {
    const styleTag = createStyleTag(targetDocument, styleTagId);
    if (styleTag) {
      insertSheet(styleTag, rule);
      return () => {
        styleTag.remove();
      };
    }
  }, [styleTagId, targetDocument, rule]);
```